Add opt for sequentializing policy_channel updates

slizco opened the pull request:
There is a known issue of parallelized policy_channel updates failing
when separate requests try to mutate the same channel. New Relic has
confirmed this happens because an update to the policy_channel link is
actually a mutation of the channel itself which includes a list of
policies. That list is not locked, so parallelized requests stomp
down on each other, leading to errors when using infrastructure
management tools, like Terraform, that fire requests in close succession.

Related issue:
https://github.com/terraform-providers/terraform-provider-newrelic/issues/100

slizco committed:
Add opt for sequentializing policy_channel updates

There is a known issue of parallelized policy_channel updates failing
when separate requests try to mutate the same channel. New Relic has
confirmed this happens because an update to the policy_channel link is
actually a mutation of the channel itself which includes a list of
policies. That list is not locked, so parallelized requests stomp
down on each other, leading to errors when using infrastructure
management tools, like Terraform, that fire requests in close succession.

Related issue:
https://github.com/terraform-providers/terraform-provider-newrelic/issues/100

slizco committed:
Add comments, change field name

paultyng commented:
I have never used `sync.Map` so perhaps I'm missing something, but it doesn't seem like these ever verifies that things are locked? It just attempts to `Store` them again or `Delete` them? I didn't see a `Load` outside of the unit testing. I wonder if the value in the map should be a `Mutex` so callers can wait for the unlocking?

slizco commented:
Right, so `sync.Map`, as I understand it, allows automatic locking for writing of specific sets of keys. I am essentially "locking" the resource each time I `Store` it in the map, because any other `Store` or `Delete` on that key has to be executed by the _same_ goroutine. Once I `Delete` the key in the same goroutine, it becomes unlocked again, and another process is free to mutate it. At least, that's my interpretation of the [godocs](https://golang.org/pkg/sync/#Map):

_"Map is like a Go map[interface{}]interface{} but is safe for concurrent use by multiple goroutines without additional locking or coordination."_

Our use-case aligns nicely with (2) here since we still want to allow concurrent updates to _different_ alert channels:

_"The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."_

I thought about including a test of the locking mechanism, as you said, but I figured that may be considered testing the `sync.Map` functions themselves. I'm happy to add another test that actually calls both `LockResources` and `UnlockResources` to ensure they are blocking concurrent executions of some wrapped code block.

slizco commented:
Hey @paultyng, I've been playing around with this, and it doesn't seem to work as expected. I've got a revision coming that should correct the issue. If I can't get updates of disjoint channel sets to operate in parallel, I'll just do the single mutex solution you suggested. Thanks!

slizco committed:
Use map of channels to achieve locking instead

Locking for each resource id is achieved by channel blocking. Each
resource has its own channel, stored in the map and accessed by
`resourceID`. Only the goroutine that has received the channel signal
can execute the wrapped code block.

slizco commented:
@paultyng, I've got a different fix in that I _think_ is working. Let me know your thoughts! Thanks!

slizco commented:
Hey @paultyng, I'm going to close this PR out in favor of https://github.com/paultyng/go-newrelic/pull/48 which uses the simple mutex solution. I've been playing around with this code a lot and still having issues getting it to work reliably in repeated tests. I think you're right, the safest thing to do is just use a standard `sync.Mutex` and force one update at a time. Let me know your thoughts! Thanks

slizco commented:
Closing in favor of https://github.com/paultyng/go-newrelic/pull/48.
